lib/fs.js

```diff
@@ -1404,34 +1406,58 @@ function mkdirSync(path, options) {
   }
 }
 
-// TODO(Ethan-Arrowood): Make this iterative too
-function readdirSyncRecursive(path, origPath, options) {
-  nullCheck(path, 'path', true);
-  const ctx = { path };
-  const result = binding.readdir(pathModule.toNamespacedPath(path),
-                                 options.encoding, !!options.withFileTypes, undefined, ctx);
-  handleErrorFromBinding(ctx);
-  return options.withFileTypes ?
-    getDirents(path, result).flatMap((dirent) => {
-      return [
-        dirent,
-        ...(dirent.isDirectory() ?
-          readdirSyncRecursive(
-            pathModule.join(path, dirent.name),
-            origPath,
-            options,
-          ) : []),
-      ];
-    }) :
-    result.flatMap((ent) => {
-      const innerPath = pathModule.join(path, ent);
-      const relativePath = pathModule.relative(origPath, innerPath);
-      const stat = binding.internalModuleStat(innerPath);
-      return [
-        relativePath,
-        ...(stat === 1 ? readdirSyncRecursive(innerPath, origPath, options) : []),
-      ];
-    });
+/**
+ * An iterative algorithm for reading the entire contents of the `basePath` directory.
+ * This function does not validate `basePath` as a directory. It is passed directly to
+ * `binding.readdir` after a `nullCheck`.
+ * @param {string} basePath
+ * @param {{ encoding: string, withFileTypes: boolean }} options
+ * @returns {string[] | Dirent[]}
+ */
+function readdirSyncRecursive(basePath, options) {
+  nullCheck(basePath, 'path', true);
+
+  const readdirResults = [];
+  const pathsQueue = [basePath];
+
+  function read(path) {
+    const ctx = { path };
+    const readdirResult = binding.readdir(
+      pathModule.toNamespacedPath(path),
+      options.encoding,
+      Boolean(options.withFileTypes),
+      undefined,
+      ctx,
+    );
+    handleErrorFromBinding(ctx);
+
+    if (options.withFileTypes) {
+      const dirents = getDirents(path, readdirResult);
+      for (let i = 0; i < dirents.length; i++) {
+        const dirent = dirents[i];
+        ArrayPrototypePush(readdirResults, dirent);
+        if (dirent.isDirectory()) {
+          ArrayPrototypePush(pathsQueue, pathModule.join(dirent.path, dirent.name));
+        }
+      }
+    } else {
+      for (let i = 0; i < readdirResult.length; i++) {
+        const resultPath = pathModule.join(path, readdirResult[i]);
+        const relativeResultPath = pathModule.relative(basePath, resultPath);
+        const stat = binding.internalModuleStat(resultPath);
+        ArrayPrototypePush(readdirResults, relativeResultPath);
+        // 1 indicates directory
+        if (stat === 1) {
+          ArrayPrototypePush(pathsQueue, resultPath);
+        }
+      }
+    }
+  }
+
+  while (pathsQueue.length > 0)
+    read(ArrayPrototypeShift(pathsQueue));
```

Comment:
This is likely way more performant.
```suggestion
  for (let i = 0; i < pathsQueue.length; i++) {
    read(pathsQueue[i]);
  }
```

Reply:
Interesting, I wouldn't have thought to do it that way. I'll add this and see if everything still passes. Thanks!